Inconsistent internal state after applying multiple slices

```python
from datatable import *
DT = Frame(A=['cat'])
DT.nrows = 3
DT = DT[2:, :]
DT.internal.check()
```
leads to
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Mismatching min/max values in the ArrayRowIndex min=0/max=0 compared to the computed min=2147483647/max=0
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -151,6 +151,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - In certain circumstances mixing computed and plain columns under groupby
   caused incorrect result (#1578).
 
+- Fixed an internal error which was occurring when multiple row filters were
+  applied to a Frame in sequence (#1592).
+
 
 ### Changed
 
@@ -217,7 +220,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
   discovering and reporting bugs that were fixed in this release:
 
   [Pasha Stetsenko][] (#1316, #1443, #1481, #1539, #1542, #1551, #1572, #1576,
-  #1578),
+  #1578, #1592),
   [Arno Candel][] (#1437, #1491, #1510, #1525, #1549, #1556, #1562),
   [Michael Frasco][] (#1448),
   [Jonathan McKinney][] (#1451, #1565),

--- a/c/expr/i_node.cc
+++ b/c/expr/i_node.cc
@@ -243,7 +243,7 @@ void slice_in::execute_grouped(workframe& wf) {
   xassert(j <= ri_size);
   out_ri_array.resize(j);
   out_groups.resize((k + 1) * sizeof(int32_t));
-  RowIndex newri(std::move(out_ri_array), /* sorted = */ true);
+  RowIndex newri(std::move(out_ri_array), /* sorted = */ (step >= 0));
   Groupby newgb(k, std::move(out_groups));
   wf.apply_rowindex(newri);
   wf.apply_groupby(newgb);

--- a/c/rowindex_array.cc
+++ b/c/rowindex_array.cc
@@ -30,6 +30,14 @@
 #include "utils/assert.h"
 #include "utils/parallel.h"
 
+static void test(ArrayRowIndexImpl* o) {
+  #ifndef NDEBUG
+    o->refcount++;
+    o->verify_integrity();
+    o->refcount--;
+  #endif
+}
+
 
 
 //------------------------------------------------------------------------------
@@ -44,6 +52,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(arr32_t&& array, bool sorted) {
   owned = array.data_owned();
   data = array.release();
   set_min_max();
+  test(this);
 }
 
 
@@ -54,6 +63,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(arr64_t&& array, bool sorted) {
   owned = array.data_owned();
   data = array.release();
   set_min_max();
+  test(this);
 }
 
 
@@ -67,6 +77,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(arr32_t&& array, size_t _min, size_t _max)
   data = array.release();
   min = _min;
   max = _max;
+  test(this);
 }
 
 
@@ -79,6 +90,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(arr64_t&& array, size_t _min, size_t _max)
   data = array.release();
   min = _min;
   max = _max;
+  test(this);
 }
 
 
@@ -144,6 +156,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(
     }
     xassert(rowsptr == static_cast<int64_t*>(data) + length);
   }
+  test(this);
 }
 
 
@@ -164,6 +177,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(const Column* col) {
     default:
       throw ValueError() << "Column is not of boolean or integer type";
   }
+  test(this);
 }
 
 
@@ -253,6 +267,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(filterfn32* ff, size_t n, bool sorted) {
   length = out_length;
   _resize_data();
   set_min_max();
+  test(this);
 }
 
 
@@ -299,6 +314,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(filterfn64* ff, size_t n, bool sorted) {
   length = out_length;
   _resize_data();
   set_min_max();
+  test(this);
 }
 
 
@@ -320,30 +336,29 @@ void ArrayRowIndexImpl::set_min_max() {
 
 template <typename T>
 void ArrayRowIndexImpl::_set_min_max() {
+  constexpr T TMAX = std::numeric_limits<T>::max();
   const T* idata = static_cast<const T*>(data);
   if (length == 1) ascending = true;
   if (length == 0) {
     min = max = RowIndex::NA;
-  } else if (length == 1) {
-    min = static_cast<size_t>(idata[0]);
-    max = static_cast<size_t>(idata[length - 1]);
-    if (min == RowIndex::NA || max == RowIndex::NA) {
-      if (min == RowIndex::NA && max == RowIndex::NA) {
-        min = max = 0;
-      } else if (min == RowIndex::NA) {
-        size_t j = 1;
-        while (j < length && idata[j] == -1) ++j;
-        min = static_cast<size_t>(idata[j]);
-      } else {
-        size_t j = length - 2;
-        while (j < length && idata[j] == -1) --j;
+  }
+  else if (ascending) {
+    for (size_t j = 0; j < length; ++j) {
+      min = static_cast<size_t>(idata[j]);
+      if (min != RowIndex::NA) break;
+    }
+    if (min == RowIndex::NA) {
+      max = min;
+    } else {
+      for (size_t j = length - 1; j < length; --j) {
         max = static_cast<size_t>(idata[j]);
+        if (max != RowIndex::NA) break;
       }
     }
-    if (min > max) std::swap(min, max);
-  } else {
-    T tmin = std::numeric_limits<T>::max();
-    T tmax = -std::numeric_limits<T>::max();
+  }
+  else {
+    T tmin = TMAX;
+    T tmax = -TMAX;
     #pragma omp parallel for schedule(static) \
         reduction(min:tmin) reduction(max:tmax)
     for (size_t j = 0; j < length; ++j) {
@@ -352,13 +367,13 @@ void ArrayRowIndexImpl::_set_min_max() {
       if (t < tmin) tmin = t;
       if (t > tmax) tmax = t;
     }
-    if (tmin == std::numeric_limits<T>::max() &&
-        tmax == -std::numeric_limits<T>::max()) tmin = tmax = -1;
+    if (tmin == TMAX && tmax == -TMAX) tmin = tmax = -1;
     min = static_cast<size_t>(tmin);
     max = static_cast<size_t>(tmax);
   }
-  xassert(max >= min && (max == RowIndex::NA || max <= RowIndex::MAX)
-                     && (min == RowIndex::NA || min <= RowIndex::MAX));
+  xassert(max >= min);
+  xassert(max == RowIndex::NA || max <= RowIndex::MAX);
+  xassert(min == RowIndex::NA || min <= RowIndex::MAX);
 }
 
 
@@ -449,7 +464,7 @@ const int64_t* ArrayRowIndexImpl::indices64() const noexcept {
 
 
 
-RowIndexImpl* ArrayRowIndexImpl::uplift_from(const RowIndexImpl* rii) {
+RowIndexImpl* ArrayRowIndexImpl::uplift_from(const RowIndexImpl* rii) const {
   RowIndexType uptype = rii->type;
   if (uptype == RowIndexType::SLICE) {
     size_t start = slice_rowindex_get_start(rii);
@@ -629,9 +644,10 @@ template <typename T>
 static void verify_integrity_helper(
     void* data, size_t len, size_t min, size_t max, bool sorted)
 {
+  constexpr T TMAX = std::numeric_limits<T>::max();
   auto ind = static_cast<const T*>(data);
-  T tmin = std::numeric_limits<T>::max();
-  T tmax = 0;
+  T tmin = TMAX;
+  T tmax = -TMAX;
   bool check_sorted = sorted;
   for (size_t i = 0; i < len; ++i) {
     T x = ind[i];
@@ -644,16 +660,16 @@ static void verify_integrity_helper(
     if (x > tmax) tmax = x;
     if (check_sorted && i > 0 && x < ind[i-1]) check_sorted = false;
   }
-  if (!len) tmin = tmax = -1;
+  if (tmin == TMAX && tmax == -TMAX) tmin = tmax = -1;
+  if (check_sorted != sorted) {
+    throw AssertionError()
+        << "ArrrayRowIndex is marked as sorted, but actually it isn't.";
+  }
   if (static_cast<size_t>(tmin) != min || static_cast<size_t>(tmax) != max) {
     throw AssertionError()
         << "Mismatching min/max values in the ArrayRowIndex min=" << min
         << "/max=" << max << " compared to the computed min=" << tmin
         << "/max=" << tmax;
-  }
-  if (check_sorted != sorted) {
-    throw AssertionError()
-        << "ArrrayRowIndex is marked as sorted, but actually it isn't.";
   }
 }
 

--- a/c/rowindex_impl.h
+++ b/c/rowindex_impl.h
@@ -36,7 +36,9 @@ class RowIndexImpl {
      *     The number of elements in the RowIndex.
      *
      * min, max
-     *     Smallest / largest entry in the RowIndex.
+     *     Smallest / largest entry in the RowIndex. If the RowIndex is empty
+     *     (length 0), or if all entries in the RowIndex are NAs, then
+     *     min = max = RowIndex::NA.
      *
      * refcount
      *     Ref-counter for this RowIndexImpl object. A RowIndexImpl* object
@@ -72,7 +74,7 @@ class RowIndexImpl {
     void release();
 
     virtual size_t nth(size_t i) const = 0;
-    virtual RowIndexImpl* uplift_from(const RowIndexImpl*) = 0;
+    virtual RowIndexImpl* uplift_from(const RowIndexImpl*) const = 0;
     virtual RowIndexImpl* negate(size_t nrows) const = 0;
 
     virtual void resize(size_t n) = 0;
@@ -97,7 +99,7 @@ class SliceRowIndexImpl : public RowIndexImpl {
     SliceRowIndexImpl(size_t start, size_t count, size_t step);
 
     size_t nth(size_t i) const override;
-    RowIndexImpl* uplift_from(const RowIndexImpl*) override;
+    RowIndexImpl* uplift_from(const RowIndexImpl*) const override;
     RowIndexImpl* negate(size_t nrows) const override;
 
     void resize(size_t n) override;
@@ -145,7 +147,7 @@ class ArrayRowIndexImpl : public RowIndexImpl {
     const int64_t* indices64() const noexcept;
 
     size_t nth(size_t i) const override;
-    RowIndexImpl* uplift_from(const RowIndexImpl*) override;
+    RowIndexImpl* uplift_from(const RowIndexImpl*) const override;
     RowIndexImpl* negate(size_t nrows) const override;
 
     void resize(size_t n) override;

--- a/c/rowindex_slice.cc
+++ b/c/rowindex_slice.cc
@@ -81,7 +81,7 @@ size_t SliceRowIndexImpl::nth(size_t i) const {
 
 
 
-RowIndexImpl* SliceRowIndexImpl::uplift_from(const RowIndexImpl* rii) {
+RowIndexImpl* SliceRowIndexImpl::uplift_from(const RowIndexImpl* rii) const {
   RowIndexType uptype = rii->type;
 
   // Product of 2 slices is again a slice.

--- a/c/set_funcs.cc
+++ b/c/set_funcs.cc
@@ -50,10 +50,10 @@ struct sort_result {
 // helper functions
 //------------------------------------------------------------------------------
 
-static py::oobj make_pyframe(
-    sort_result& sr, arr32_t&& arr, bool arr_sorted = true
-) {
-  RowIndex out_ri = RowIndex(std::move(arr), arr_sorted);
+static py::oobj make_pyframe(sort_result& sr, arr32_t&& arr) {
+  // The array of rowindices `arr` is typically shuffled because the values
+  // in the input are sorted before they are compared.
+  RowIndex out_ri = RowIndex(std::move(arr), false);
   Column* out_col = sr.col->shallowcopy(out_ri);
   out_col->reify();
   DataTable* dt = new DataTable({out_col}, {sr.colname});
@@ -420,8 +420,7 @@ static py::oobj _symdiff(ccolvec&& cc) {
     }
   }
   arr.resize(j);
-  // symdiff() is the only method that can produce potentially unsorted `arr`.
-  return make_pyframe(sr, std::move(arr), false);
+  return make_pyframe(sr, std::move(arr));
 }
 
 

--- a/tests/munging/test_dt_rows.py
+++ b/tests/munging/test_dt_rows.py
@@ -219,6 +219,16 @@ def test_slice_errors2(dt0):
                       "must be positive")
 
 
+def test_slice_after_resize():
+    """Issue #1592"""
+    DT = dt.Frame(A=['cat'])
+    DT.nrows = 3
+    RES = DT[2:, :]
+    RES.internal.check()
+    assert RES.to_list() == [[None]]
+
+
+
 
 #-------------------------------------------------------------------------------
 # Range selector

--- a/tests/random_attack.py
+++ b/tests/random_attack.py
@@ -313,10 +313,7 @@ class Frame0:
     def resize_rows(self, nrows):
         curr_nrows = self.nrows
         self.df.nrows = nrows
-        if curr_nrows == 1:
-            for i, elem in enumerate(self.data):
-                self.data[i] = elem * nrows
-        elif curr_nrows < nrows:
+        if curr_nrows < nrows:
             append = [None] * (nrows - curr_nrows)
             for i, elem in enumerate(self.data):
                 self.data[i] = elem + append

